servlet/src/main/java/io/undertow/servlet/api/WebXMLParser.java

```diff
@@ -0,0 +1,629 @@
+package io.undertow.servlet.api;
+
+import io.undertow.jsp.JspFileHandler;
+import io.undertow.jsp.JspServletBuilder;
+import io.undertow.servlet.core.ManagedServlet;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.EventListener;
+import java.util.HashMap;
+import java.util.LinkedList;
+import javax.servlet.DispatcherType;
+import javax.servlet.Filter;
+import javax.servlet.MultipartConfigElement;
+import javax.servlet.Servlet;
+import javax.servlet.descriptor.JspConfigDescriptor;
+import javax.servlet.descriptor.JspPropertyGroupDescriptor;
+import javax.servlet.descriptor.TaglibDescriptor;
+import javax.servlet.http.HttpServlet;
+import javax.xml.parsers.DocumentBuilder;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import org.apache.jasper.runtime.InstanceManagerFactory;
+import org.apache.jasper.servlet.JspServlet;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+
+public class WebXMLParser {
+    protected interface WebXMLParserFactory {
+        void parse(DeploymentInfo deploymentInfo, Element element) throws ClassNotFoundException, WebXMLException;
```

Comment:
I had commented the choice in DOM, but could look into STAX if it's really desired. The problem with STAX is that it completely ignores the XML structure, only sending tag names regardless of depth. It also requires storing a lot of state whereas the current DOM implementation stores almost all state in the DeploymentInfo instance that is the end result. The current implementation could technically parse multiple web.xml files simultaneously, although that really isn't the intention of the servlet spec. JAXB would be the best of both worlds, but would require an extra dependency to be added for the implementation library which I wanted to avoid. Let me know how I should proceed.

>            //using DOM parser since:
            // a) web.xml shouldn't be large enough to not be loadable into memory
 +            // b) web.xml should only need to be parsed once at startup, so the memory used by DOM parser will be freed
            // c) the structure of web.xml will be much easier to handle with DOM as opposed to SAX
            // d) using JAXB, while even simpler than DOM, would add an additional dependency to the project